src/v/storage/tests/segment_concatenation_test.cc

```diff
@@ -0,0 +1,196 @@
+};
+
+segment_comparision_fields concat_read_segments(
+  std::vector<ss::lw_shared_ptr<storage::segment>> segments) {
+    size_t start_pos = 0;
+    size_t end_pos = segments.back()->file_size();
+    storage::concat_segment_reader_view cv{
+      segments, start_pos, end_pos, ss::default_priority_class()};
+    iobuf buf;
+    segment_comparision_fields res;
+    auto transform_stream_result
+      = storage::transform_stream(
+          cv.take_stream(),
+          make_iobuf_ref_output_stream(buf),
+          [&res](model::record_batch_header h) {
+              res.num_records += h.record_count;
+              res.size_bytes += h.record_count;
+              return storage::batch_consumer::consume_result::accept_batch;
+          })
+          .get();
+
+    res.base_offset = segments.front()->offsets().get_base_offset();
+    res.dirty_offset = segments.back()->offsets().get_dirty_offset();
+    res.committed_offset = segments.back()->offsets().get_committed_offset();
+    res.stable_offset = segments.back()->offsets().get_stable_offset();
+    EXPECT_TRUE(transform_stream_result.has_value());
+    res.read_result = read_segments_result{
+      .buf = std::move(buf), .len = transform_stream_result.value()};
+    return res;
+}
+
+void concatenate_segments_from_log(
+  storage::disk_log_impl& log,
+  int expected_segments_to_compact,
+  ss::sharded<features::feature_table>& feature_table) {
+    storage::compaction_config cfg(
+      model::offset::max(),
+      std::nullopt,
+      ss::default_priority_class(),
+      never_abort);
+
+    const auto closed_segment_filter = [](const auto& s) -> bool {
+        return !s->has_appender();
+    };
+
+    auto log_segments = log.segments()
+                        | std::views::filter(closed_segment_filter);
+
+    std::vector<ss::lw_shared_ptr<storage::segment>> segments(
+      log_segments.begin(), log_segments.end());
+    EXPECT_EQ(segments.size(), expected_segments_to_compact);
+
+    // Segments must be self compacted before they can be concatenated.
+    for (auto& seg : segments) {
+        log.segment_self_compact(cfg, seg).get();
+    }
+
+    auto expected_target_file_size = std::accumulate(
+      segments.begin(),
+      segments.end(),
+      size_t{0},
+      [](size_t sum, const auto& seg) { return sum + seg->file_size(); });
+
+    auto segments_read_result = concat_read_segments(segments);
+
+    // the segment which will be expanded to replace
+    auto target = segments.front();
+
+    [[maybe_unused]] auto ret
+      = storage::internal::concatenate_and_rebuild_target_segment(
+          target,
+          segments,
+          log.stm_manager(),
+          cfg,
+          log.get_probe(),
+          log.readers(),
+          log.resources(),
+          feature_table)
+          .get();
+
+    EXPECT_EQ(target->file_size(), expected_target_file_size);
+
+    auto target_read_result = concat_read_segments({target});
+    EXPECT_EQ(segments_read_result, target_read_result);
```

Comment:
WDYT about adding a test that exercises the time between concating and removal, when the log has some extra data? You might have a test that covers it later, but wondering if it's easy to have an explicit test for that case here, given we're using the compaction primitives in this test

Reply:
It does, but I generally think it's a great idea to have a test that does the test in a more direct, simpler way than triggering exceptions.